File: 4.semester/IJA/src/application/Port.java
/* 
 * Project: 	Implementing Block Editor in JavaFX
 * Course: 		IJA 2018
 * Team name: 	Tym xzubri00
 * Members: 
 *		Liscinsky Matus	xlisci02
 *		Zubrik Tomas 	xzubri00
 */
package application;
import java.util.Vector;
import java.util.HashMap;
import java.util.Map;

public class Port {
	HashMap <String, Double> m;
	Vector <String> keys;
	Port connected_to;
	Block in_block ;

	/**
	 * Constructor of Port
	 */
	public Port() {
		this.connected_to=null;
		m = new HashMap <String,Double> ();
		keys = new Vector <String> ();
	}

	/**
	 * Function sets complex number on port
	 * @param real      real part
	 * @param imaginary imaginary part
	 */
	void setComplex(double real, double imaginary) {
		this.setReal(real);
		this.setImaginary(imaginary);
	}
	 
	/**
	 * Function adds type key in vector of keys 
	 * @param key Key to be added in vector
	 */
	void addTypeKey(String key) {
		this.keys.addElement(key);
	}
	 
	/**
	 * Function sets real part of complex number on port
	 * @param real Real part of complex number
	 */
	void setReal(double real) {
		this.m.put("real", real);
	}
	 
	/**
	 * Function sets imaginary part of complex number on port
	 * @param imaginary Imaginary part of complex number
	 */
	void setImaginary(double imaginary) {
		this.m.put("im", imaginary);
	}
	 
	/**
	 * Function sets complex number in polar form on port
	 * @param magnitude   magnitude of complex number
	 * @param phase_angle phase angle of complex number
	 */
	void setPolar(double magnitude, double phase_angle) {
		this.m.put("magnitude", magnitude);
		this.m.put("angle", phase_angle);
	}
	 
	/**
	 * Function gets information about map of port
	 * @return Formatted String with information about map
	 */
	String mapToString(){
		String result = new String();
		    if (this.m.isEmpty()){
		        result = result.concat("empty");
		        return result;
		    }
		    for (Map.Entry<String, Double> entry : this.m.entrySet()) {
		    	String key = entry.getKey();
		        Double value = entry.getValue();
		        
		        result = result.concat("  ");
		        result = result.concat(key);
		        result = result.concat(": ");
		        result = result.concat(value.toString());
		        result = result.concat("\n");
		    }
		    return result.substring(0, result.length()-1);
	}
}

File: 4.semester/ICP/src/port.cpp
/**
* @file port.cpp
* @brief Port implementation
* @author Matúš Liščinský xlisci02, Marcin Vladimír xmarci10
*/
#include "port.h"

/**
 * @brief Port constructor
 */
Port::Port()
{
    this->connected_to=NULL;
}

/**
 * @brief Instert complex number to map
 * @param real real part of complex number
 * @param imaginary imaginary part of complex number
 */
void Port::setComplex(double real, double imaginary){
    this->setReal(real);
    this->setImaginary(imaginary);
}

/**
 * @brief Add new item to port type
 * @param new item
 */
void Port::addTypeKey(std::string key){
    this->keys.push_back(key);
}

/**
 * @brief Insert real part of complex number to map
 * @param real part which will be set
 */
void Port::setReal(double real){
    this->m.insert(std::make_pair("real", real));
}

/**
 * @brief Insert imaginary part of complex number to map
 * @param imaginary part which will be set
 */
void Port::setImaginary(double imaginary){
    this->m.insert(std::make_pair("im", imaginary));
}

/**
 * @brief Set polar form of complex number to map
 * @param magnitude magnitude part of polar form
 * @param phase_angle phase part of polar form
 */
void Port::setPolar(double magnitude, double phase_angle){
    this->m.insert(std::make_pair("magnitude", magnitude));
    this->m.insert(std::make_pair("angle", phase_angle));
}

/**
 * @brief Convert map contain to string
 * @return return created string
 */
std::string Port::mapToString(){
    std::string result;
    if(this->m.empty()){
        for(auto &x : this->keys){
            result.append("  ");
            result.append(x);
            result.append(": -\n");
        }
        return result.substr(0,result.size()-1);
    }
    for(auto &x : this->m){
        result.append("  ");
        result.append(x.first);
        result.append(": ");
        // zaokruhlenie
        std::stringstream stream;
        stream << std::fixed << std::setprecision(2) << x.second;
        std::string val = stream.str();
        result.append(val);
        result.append("\n");
    }
    return result.substr(0,result.size()-1);
}
